### ISSUE TITLE
useMutation onError handler is not fired if error policy is set to "all"

### ISSUE THREAD

**Issue by Imran99** [apollographql/apollo-client#10973]
### Intended Outcome
From the [docs](https://www.apollographql.com/docs/react/data/mutations/#usemutation-api) the onError policy is called unless the `errorPolicy` is set to `ignore`.

### Actual Outcome
useMutation onError handler is not called when an error occurs on the graphql server and the error policy is set to `all`.

### Versions
`"@apollo/client": "^3.7.15"`

### Link to Reproduction

https://codesandbox.io/s/thirsty-darkness-crw7f8?file=/src/index.jsx

### Reproduction Steps

1. Invoke a mutation with an onError handler and errorPolicy: "all"
2. When an error occurs the onError handler will not be called.

The code sandbox link reproduces the issue.

**Comment by CH-PatrickWalmsley** [comment]
+1 

Heads up to others - onComplete does not adhere to the behavior described in the docs either.  🤕 

As a potential workaround you can use update on place of onError and onComplete.  First param is a proxy to the cache.  Second param is the raw response with both data and errors.  Unlike onError and onComplete, update will not run after the component scope variables are updated so you must use the response contents passed to the method.

**Comment by bignimbus** [comment]
Hi @Imran99 👋🏻 I'm looking at the reproduction and don't see where the error is being thrown that would trigger `onError`.  When I add a rejected promise to the link [in this fork](https://codesandbox.io/s/aged-river-jnxxmc), I see `onError was called` in the console upon executing the mutation.  Apologies if I'm missing something obvious!

@CH-PatrickWalmsley let me know more about what you're seeing, do you have something runnable to take a look at?  We do have a suite of unit tests that validate callback behavior, though I admit that sometimes they don't behave according to folks' expectations and are considering changing our approach in an upcoming major version.

Eager to hear from either/both of you on how to figure things out from here 🙏🏻

**Comment by Imran99** [comment]
Apologies @bignimbus I think I may have had some unsaved changes in my fork. I've added the throw back in which reproduces the issue: https://codesandbox.io/s/thirsty-darkness-crw7f8?file=/src/index.jsx

<img width="1775" alt="Screen Shot 2023-06-29 at 17 36 32" src="https://github.com/apollographql/apollo-client/assets/2307325/4359729d-e340-4a8f-b75e-c80bc7238672">

**Comment by chawomack** [comment]
+1

I'm noticing that `onError` does not get executed and the `error` field remains undefined.

**Comment by bignimbus** [comment]
Ok thanks for that, I'll share this with the team and get some more context.  From what I understand, not every Error in your application can/should be expected to be handled through Apollo Client.  Network errors and spec-compliant GraphQL errors from the server will be reported, but not necessarily client-side runtime errors.  I'd categorize your example, which throws a plain JavaScript error inside the resolver of the mutation type, as neither of those.  If such an error were thrown on the server it would either be caught and passed down in the `error `payload or reported as a `5xx` network error.  If you look at the fork I shared above, you'll notice the error is passed to `onError` without issue.

Can you share a little more about what you're seeing in the application you're working on?  Where is the error being thrown?

**Comment by Imran99** [comment]
Thanks @bignimbus. At the moment the error handling behaviour seems to be a bit inconsistent. If you set errorPolicy to `none`, `onError is called`. If you set errorPolicy to `all`, `onError isn't called`. I was expecting any error in the error response collection to trigger `onError` as this made more sense when considered with client errorPolicy settings.

<img width="2551" alt="image" src="https://github.com/apollographql/apollo-client/assets/2307325/9c13ccc2-75a6-468f-8c16-58a9d1c0f001">

> Can you share a little more about what you're seeing in the application you're working on? Where is the error being thrown?
My scenario is similar to the one in the sample application. I have some unexpected error happening in the resolver that I want to handle with a generic oops message on a particular screen that makes the request.

**Comment by bignimbus** [comment]
> I have some unexpected error happening in the resolver

Is this a resolver on the server, a local resolver, or using a `graphql.js` resolver directly?

**Comment by Imran99** [comment]
irl the resolver is on the server yup. The behaviour of onError is the same.

**Comment by bignimbus** [comment]
Ok thanks 🙏🏻 are you able to paste the JSON returned from the server?  Feel free to remove any sensitive data, I'm more curious to hear whether the response has an `"errors"` property that [conforms to the spec](https://spec.graphql.org/October2021/#sec-Errors)

**Comment by Imran99** [comment]
Ok i'll see if I can dig that up. Note  I'm using apollo server on the server side and that is handling the error formatting.

**Comment by Imran99** [comment]
Sample error:

```js
  [
    {
      "errors": [
        {
          "message": "Request failed with status code 400",
          "locations": [
            {
              "line": 5,
              "column": 9
            }
          ],
          "path": [
            "orders",
            "nodes",
            0,
            "fulfillmentOrders",
            0,
            "quotes"
          ],
          "extensions": {
            "code": "INTERNAL_SERVER_ERROR",
            "stacktrace": [
              "AxiosError: Request failed with status code 400",
              "...obfuscated"
            ]
          }
        }
      ],
      "data": {
        "orders": {
          "nodes": [
            {
              "fulfillmentOrders": [
                {
                  "quotes": null,
                  "__typename": "FulfillmentOrder"
                }
              ],
              "__typename": "Order"
            }
          ],
          "__typename": "OrderConnection"
        }
      }
    }
  ]
```

**Comment by bignimbus** [comment]
Thanks for this info, it's much appreciated and helps narrow things down.  I'm going to see if I can reproduce this in the fork and will get back to you!

**Comment by jerelmiller** [comment]
@bignimbus looks like you're using an `errorPolicy` of `none`. I'm able to reproduce if you change `errorPolicy` to `all`. `onCompleted` is called instead in that case.

**Comment by bignimbus** [comment]
Ah my mistake, thanks for catching that @jerelmiller

**Comment by milesingrams** [comment]
Dealing with the same. `errorPolicy: all` and inline `onError` is not called in `useQuery`.

**Comment by jerelmiller** [comment]
This has been fixed in #11103 and will be released with 3.8.0 in the next couple weeks. Thanks for reporting this!

**Comment by Imran99** [comment]
Thanks @jerelmiller !
